### PR TITLE
[WIP] NHWC support when running with MKL-DNN

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

NHWC support when running with MKL-DNN

## How was this patch tested?

unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRElement.scala [hunk 1]
```diff
@@ -85,8 +85,8 @@ case class IRSpatialBatchNormalization[T: ClassTag](
 
 case class IRIdentity[T: ClassTag]() extends IROperator[T]
 
-case class IRDropout[T: ClassTag](initP: Double = 0.5, inplace: Boolean = false,
-                                  scale: Boolean = true) extends IROperator[T]
+//case class IRDropout[T: ClassTag](initP: Double = 0.5, inplace: Boolean = false,
```
Comment (benjamim93) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRElement.scala]: Need to insert a space after the start of the comment

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasLayer.scala [hunk 2]
```diff
@@ -253,10 +295,19 @@ abstract class KerasLayer[A <: Activity: ClassTag, B <: Activity: ClassTag, T: C
...
     if (!nodes.isEmpty) {
-    val inputShape = Shape(nodes.map{_.element.getOutputShape()}.toList)
-      this.build(inputShape)
+      // Also, no need to validate the input
+      val inputShape = Shape(nodes.map{_.element.getOutputShape()}.toList)
+      if (!this.isBuilt()) {
+        this.build(inputShape)
+        validateInput(nodes.map(_.element))
+
+        if (!nodes.isEmpty) {
```
Comment (benjamim93) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasLayer.scala]: Here might be some mistake.

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/keras/nn/KerasStyleSpec.scala [hunk 3]
```diff
@@ -272,4 +273,64 @@ class KerasStyleSpec extends BigDLSpecHelper {
...
+    // graph.evaluate()
+    val ir = graph.asInstanceOf[StaticGraph[Float]].cloneModule().toIRgraph()
+    val tensor = Tensor[Float](Array(3, 10)).rand()
+
+    val outputBlas = graph.forward(tensor)
+    val output = ir.forward(tensor)
+
+    val gradOutput = Tensor[Float]().resizeAs(outputBlas.toTensor[Float]).rand()
+    val gradInputBlas = graph.backward(tensor, gradOutput)
+    val gradInput = graph.backward(tensor, gradOutput)
```
Comment (benjamim93) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/keras/nn/KerasStyleSpec.scala]: Is it ir.bakward()?

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/keras/nn/KerasStyleSpec.scala [hunk 4]
```diff
@@ -272,4 +273,64 @@ class KerasStyleSpec extends BigDLSpecHelper {
...
+    val ir = graph.asInstanceOf[StaticGraph[Float]].cloneModule().toIRgraph()
+
+    val tensor = Tensor[Float](Array(3, 20, 100)).rand()
+    val outputBlas = graph.forward(tensor)
+    val output = ir.forward(tensor)
+    outputBlas should be(output)
+
+    val gradOutput = Tensor[Float]().resizeAs(outputBlas.toTensor[Float]).rand()
+    val gradInputBlas = graph.backward(tensor, gradOutput)
+    val gradInput = graph.backward(tensor, gradOutput)
```
Comment (benjamim93) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/keras/nn/KerasStyleSpec.scala]: Is it ir.backward()?

**Comment by zhangxiaoli73** [comment]
open a new pr, so close this one.

**Pull request closed without merging**